### PR TITLE
C#: Add local filesystem writes as External Location sinks

### PR DESCRIPTION
**Pull request by joefarebrother**
Adds sinks for local filesystem writes to `ExternalLocationSink` (used by the `cs/cleartext-storage` and `cs/exposure-of-sensitive-information` queries).

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll [hunk 1]
```diff
@@ -63,3 +64,68 @@ class CookieStorageSink extends ExternalLocationSink, RemoteFlowSink {
     )
   }
 }
+
+private predicate isFileWriteCall(Expr stream, Expr data) {
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:68]: Would it be possible to use Models as Data instead (maybe define a new sink kind if needed)?

Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:68]: Possibly; however since we care about two different parameters (the stream and the written data) then we'd need to encode that information in the sink kind; similar to [regex use sinks](https://github.com/github/codeql/blob/main/java/ql/lib/ext/java.lang.model.yml#L24) in Java [are used](https://github.com/github/codeql/blob/main/java/ql/lib/semmle/code/java/regex/RegexFlowConfigs.qll#L21) by the ReDoS queries. Is this worthwhile?

The sources for this config could also more straigtforwardly be converted to MaD.

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:68]: If it add extra complexity just leave it as it is - unless you prefer to do otherwise.
If the sources are easily convertible to MaD then it is probably a good idea (this will be relevant for the high level configuration prototype for including *more* (all sources)) when running a query.

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll [hunk 2]
```diff
@@ -63,3 +64,55 @@ class CookieStorageSink extends ExternalLocationSink, RemoteFlowSink {
...
+    stream = mc.getQualifier() and
+    data = mc.getArgument(0)
+    or
+    m.hasQualifiedName("System.Xml.Linq", "XDocument", ["Save", "SaveAsync"]) and
+    data = mc.getQualifier() and
+    stream = mc.getArgument(0)
+  )
+}
+
+private module LocalFileOutputStreamConfig implements DataFlow::ConfigSig {
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:85]: To make sure that I understand what is going on 😄 :
This configuration is used for tracking flow from the "open" of a file write stream (or file) to the "actual" write to the stream or file?
It would be great, if you could add a small code comment even though it is a private module.

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll [hunk 3]
```diff
@@ -63,3 +64,55 @@ class CookieStorageSink extends ExternalLocationSink, RemoteFlowSink {
...
+  predicate isSink(DataFlow::Node sink) { isFileWriteCall(sink.asExpr(), _) }
+
+  predicate isBarrier(DataFlow::Node node) {
+    node.asExpr()
+        .(ObjectCreation)
+        .getObjectType()
+        .hasQualifiedName("System.Security.Cryptography", "CryptoStream")
+  }
+
+  predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:97]: Why is this needed? Would it be possible to model this as summary (propagation via the StreamWriter constructor) or is it because you are making a dataflow configuration and such a summary is probably of "taint" kind?

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll [hunk 4]
```diff
@@ -63,3 +64,55 @@ class CookieStorageSink extends ExternalLocationSink, RemoteFlowSink {
...
+
+private module LocalFileOutputStreamFlow = DataFlow::Global<LocalFileOutputStreamConfig>;
+
+/**
+ * A write to the local filesystem.
+ */
+class LocalFileOutputSink extends ExternalLocationSink {
+  LocalFileOutputSink() {
+    exists(DataFlow::Node streamSink |
+      LocalFileOutputStreamFlow::flow(_, streamSink) and
```
Comment (michaelnebel) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:114]: You can use `flowTo` instead.

**Review by michaelnebel — COMMENTED** [review]
This involves some learning for me - so I have added a couple of clarifying questions 😄

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll (hunk 3 above)
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:97]: Yes; this configuration is not a taint-tracking one since most taint steps wouldn't be relevant. 
It could possibly be made a taint-tracking config for a minor precision loss (though it may not be very likely that unrelated taint steps still end up at a sink) and then this step could be made into a taint tracking MaD row (which it doesn't appear to currently be in full generality).

**Review by michaelnebel — APPROVED** [review]
LGTM!

**Comment by michaelnebel** [comment]
Great work @joefarebrother ! Thank you!

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll (hunk 3 above)
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/dataflow/flowsinks/ExternalLocationSink.qll:98]: Alright - understood! Makes sense to keep it as is!
